uefi/src/data_types/strs.rs

```diff
@@ -389,6 +402,32 @@ impl CStr16 {
         }
         Ok(())
     }
+
+    /// Like [`to_bytes_with_nul`] but without the terminating null character.
+    #[must_use]
+    pub fn to_bytes(&self) -> &[u8] {
+        let slice = self.to_bytes_with_nul();
+        &slice[..slice.len() - 2]
+    }
+
+    /// Returns the underlying bytes as slice including the terminating null
+    /// character.
+    #[must_use]
+    pub fn to_bytes_with_nul(&self) -> &[u8] {
+        unsafe { slice::from_raw_parts(self.0.as_ptr().cast(), self.num_bytes()) }
+    }
+}
+
+impl AsRef<[u8]> for CStr16 {
+    fn as_ref(&self) -> &[u8] {
+        self.to_bytes_with_nul()
+    }
+}
+
+impl Borrow<[u8]> for CStr16 {
+    fn borrow(&self) -> &[u8] {
+        self.to_bytes_with_nul()
+    }
```

Comment:
I think it would be better to leave out the AsRef/Borrow impls, as it would not be as clear when reading the code as an explicit call to `to_bytes_with_nul`.

Reply:
I'd much rather drop `self.to_bytes()` and rename `self.to_bytes_with_nul()` to `self.to_bytes()`. Then we can keep AsRef and Borrow. What do you think?

Reply:
Yeah that sounds reasonable. Also, I didn't think of this before, but it should probably be `as_bytes` instead of `to_bytes` (per the [conversion naming guidelines](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv)).

For Borrow and AsRef, do you have an example of how they'd get used in practice? Like `<CStr16 as AsRef<[u8]>>::as_ref(string)` is pretty verbose so I assume a user would usually prefer to write `string.as_bytes()`. But perhaps there's some generic code where the former would be more usable? An example would help me clear up my doubts since I always get a bit confused with those two very-similar-seeming traits.

Reply:
Sure. For example `FileSystem::write()` consumes `content: impl AsRef<[u8]>` as second parameter. This is equivalent to the standard lib's implementation.

I only implemented both, AsRef and Borrow, as the standard library also does this for a lot of types. In my past years I've came accros a lot of code that used `AsRef`-types as parameters but `Borrow` only once or twice tho. Probably good to have both, still.

Reply:
**Update** Hm, now I'm confused.. <https://doc.rust-lang.org/core/ffi/struct.CStr.html> uses `to_bytes()` and `to_bytes_with_nul()` methods, but those method implementations are cheap.

But I do not see a strong reason why we should be consistent with  <https://doc.rust-lang.org/core/ffi/struct.CStr.html>

Reply:
Interesting. I would guess the reason is this:

> Note: This method is currently implemented as a constant-time cast, but it is planned to alter its definition in the future to perform the length calculation whenever this method is called.

Several `CStr` methods make note that the implementation of the type may change in the future to make it more FFI friendly. Not sure how likely that is to ever actually happen.

I'm realizing I don't have a very strong opinion here :)

How about: just provide the `as_bytes` method (so no separate with/without null methods), and also include the `AsRef` and `Borrow` impls?

Reply:
Done